lambdas/data-migration2/src/executions.ts

```diff
@@ -0,0 +1,108 @@
+import Knex from 'knex';
+import Logger from '@cumulus/logger';
+
+import DynamoDbSearchQueue from '@cumulus/aws-client/DynamoDbSearchQueue';
+import { envUtils } from '@cumulus/common';
+import { ExecutionRecord } from '@cumulus/types/api/executions';
+import { PostgresExecutionRecord, translateApiExecutionToPostgresExecution } from '@cumulus/db';
+import { RecordAlreadyMigrated } from './errors';
+import { MigrationSummary } from './types';
+
+const Manager = require('@cumulus/api/models/base');
+const schemas = require('@cumulus/api/models/schemas');
+const Execution = require('@cumulus/api/models/executions');
+
+const logger = new Logger({ sender: '@cumulus/data-migration/executions' });
+
+/**
+ * Migrate execution record from Dynamo to RDS.
+ *
+ * @param {AWS.DynamoDB.DocumentClient.AttributeMap} dynamoRecord
+ *   Source record from DynamoDB
+ * @param {Knex} knex - Knex client for writing to RDS database
+ * @returns {Promise<number>} - Cumulus ID for record
+ * @throws {RecordAlreadyMigrated}
+ *   if record was already migrated
+ */
+export const migrateExecutionRecord = async (
+  dynamoRecord: AWS.DynamoDB.DocumentClient.AttributeMap,
+  knex: Knex
+): Promise<number> => {
+  // Use API model schema to validate record before processing
+  Manager.recordIsValid(dynamoRecord, schemas.execution);
+
+  const existingRecord = await knex<PostgresExecutionRecord>('executions')
+    .where({
+      arn: dynamoRecord.arn,
+    })
+    .first();
+  // Throw error if it was already migrated.
+  if (existingRecord) {
+    throw new RecordAlreadyMigrated(`Execution arn ${dynamoRecord.arn} was already migrated, skipping`);
+  }
+
+  const updatedRecord = await translateApiExecutionToPostgresExecution(
+    <ExecutionRecord>dynamoRecord, <Knex>knex
```

Comment:
I'm interested why the `<Knex>knex` is necessary? If the type of the variable is already set to Knex in the arguments on line 29?

Similarly, can we set the type of `dynamoRecord` to `ExecutionRecord` in the function signature or does that not work?

Reply:
Yeah, so the `<Knex>knex` isn't really necessary. I had it that way just because the other argument, `<ExecutionRecord>dynamoRecord` has that type casting and thought it looked more consistent with both arguments set up that way.

The `dynamoRecord` argument _does_ need to be cast as `ExecutionRecord` but we can definitely do it earlier than this call. My latest commit does what you're suggesting in updating the function signature to:

```
export const migrateExecutionRecord = async (
  dynamoRecord: ExecutionRecord,
  knex: Knex
): Promise<number> => { ... }
```
And removing the types from the call to the translate function:

```
  const updatedRecord = await translateApiExecutionToPostgresExecution(
    dynamoRecord, knex
  );
```

In order to make that work we _do_ need to cast `dynamoRecord` as an `ExecutionRecord` because it's being created as an AttributeMap in `migrateExecutions` in this line:

`let record = await searchQueue.peek();`

I think it makes sense to just move the cast to ExecutionRecord down to just after that happens on line 87:

```
try {
      await migrateExecutionRecord(<ExecutionRecord>record, knex);
      migrationSummary.success += 1;
    } catch (error) { ... }
```

tl;dr: We do need to cast the `dynamoRecord` as an `ExecutionRecord` but we can do it earlier and make the changes you're suggesting.